### PR TITLE
Add support for converting errors wrapped by TracedError

### PR DESCRIPTION
**Pull request by yaahc**
## Motivation

Right now TracedError is pretty much incompatible with enum heavy error
handling implementations. The way its currently designed you'd have to make
sure that each leaf error independently is wrapped in a `TracedError` so that
all sources end up capturing a SpanTrace.

## Solution

To resolve this I've added a `map` method for transforming the inner type of a
TracedError while preserving the existing SpanTrace. This is still a bit
cumbersome, where before you'd be able to write `result.map_err(|source|
MyError { source, some_context })` now you'll have to write
`result.map_err(|source| source.map(|source| MyError { source, some_context
}))`. To fix this we should probably eventually add an extension trait on
`Result` but I want to experiment with that half of the change in `zebra`
before committing to an API, since it can easily be done out of tree.

### CODE REVIEW

File: tracing-error/Cargo.toml
Comment (yaahc) [inline tracing-error/Cargo.toml:11]:
```suggestion
version = "0.2.0"
```

File: tracing-error/src/error.rs
Comment (hawkw) [inline tracing-error/src/error.rs:68]: nit:
```suggestion
    /// attached `SpanTrace`.
```

File: tracing-error/src/error.rs
Comment (hawkw) [inline tracing-error/src/error.rs:101]: nit, take it or leave it: since the return type doesn't depend on the actual _type_ of the mapped function (it's applied immediately), we could just use `impl FnOnce` here, which might be a little more concise?
```suggestion
    pub fn map<F>(self, op: impl FnOnce(E) -> F) -> TracedError<F>
    where
```

File: tracing-error/src/error.rs
Comment (hawkw) [inline tracing-error/src/error.rs:113]: nit:
```suggestion
        // specified it, so that we can safely access the vtable and spantrace fields through a type
```

File: tracing-error/src/error.rs
Comment (hawkw) [inline tracing-error/src/error.rs:128]: take it or leave it: this is quite similar to the code & safety comment here https://github.com/tokio-rs/tracing/blob/e2fa7caa12f7b1c4d00f31028bb817b0ee736f48/tracing-error/src/error.rs#L137-L158

except that the span trace comes from a different place. I wonder if it's possible to abstract over that with a private function that constructs an `ErrorImpl` for an inner error value and a provided span trace, ensuring that the vtable is valid? that way, we would put the safety-critical code in one place.

if it's not possible for such a function to ensure that the constructed `ErrorImpl` is sound, it's fine having this at each callsite, just a thought.

**Review by hawkw — APPROVED** [review]
This looks good to me. I had some non-blocking suggestions for style etc.

As a side note: I wonder if this makes it possible to add an
```rust
impl<E, F> Into<TracedError<F>> for TracedError<E>
where
    E: Into<F>,
{
    fn into(self) -> TracedError<F> {
        self.map(|e| e.into())
    }
}
```

If that works, it could be pretty handy...

**Comment by yaahc** [comment]
> As a side note: I wonder if this makes it possible to add an

Sadly no, that overlaps with `From<T> for T`, so it is crimes :(

File: tracing-error/src/error.rs
Comment (yaahc) [reply tracing-error/src/error.rs:101]: I'm a little worried about using `impl Trait` syntax here because it disables turbofish, which might be useful when dealing with type inference

File: tracing-error/src/error.rs
Comment (yaahc) [reply tracing-error/src/error.rs]: sounds like a good idea

**Comment by hawkw** [comment]
> > As a side note: I wonder if this makes it possible to add an
> 
> Sadly no, that overlaps with `From<T> for T`, so it is crimes :(

ugh. well, we could add a convenience method like `err_into`, I guess? could be a follow-up

File: tracing-error/src/error.rs
Comment (hawkw) [reply tracing-error/src/error.rs]: @yaahc did you want to do this or shall i just go ahead and merge this?

Comment (yaahc) [reply tracing-error/src/error.rs]: I want to do this before the merge

File: tracing-error/src/error.rs
Comment (yaahc) [inline tracing-error/src/error.rs:139]: I'm wondering if we could just call this `into`, I think method resolution might work out in our favor here, I'm trying to think of scenarios where it could break...

Comment (yaahc) [reply tracing-error/src/error.rs:139]: I created an example and tested it out and it seems to work great, but I don't think there's much gained, and this at least avoids the issue and any potential confusion, so I think we should leave it as is.

File: tracing-error/src/error.rs
Comment (hawkw) [inline tracing-error/src/error.rs:138]: an example would be nice here?

Comment (hawkw) [inline tracing-error/src/error.rs:138]: might reword this
```suggestion
    /// Convert the inner error type of a `TracedError` using the inner error's `Into`
    /// implementation, while preserving the attached `SpanTrace`.
```

**Comment by yaahc** [comment]
I'm not sure why the CI failed, so I restarted it

File: examples/examples/map-traced-error.rs
Comment (hawkw) [inline examples/examples/map-traced-error.rs:8]: It looks like the example needs to be updated to track that `ErrorLayer` was renamed on master:
```suggestion
use tracing_error::{prelude::*, ErrorSubscriber, TracedError};
```

File: examples/examples/map-traced-error.rs
Comment (hawkw) [inline examples/examples/map-traced-error.rs:13]:
```suggestion
        .with(tracing_subscriber::fmt::subscriber())
```

File: examples/examples/map-traced-error.rs
Comment (hawkw) [inline examples/examples/map-traced-error.rs:15]:
```suggestion
        // The `ErrorSubscriber` subscriber enables the use of `SpanTrace`.
        .with(ErrorSubscriber::default())
```

**Review by hawkw — CHANGES_REQUESTED** [review]
Looks like this needs to be updated to track the type renaming on master. Should be good to merge once that's taken care of!

File: tracing-error/src/error.rs
Comment (hawkw) [reply tracing-error/src/error.rs]: (fwiw, when i said "an example would be nice", I was thinking just like, a small API usage example _here_, rather than a full-on `examples/` example...but that's also fine! :) )

**Review by hawkw — APPROVED** [review]

**Comment by BatmanAoD** [comment]
@yaahc @hawkw Did this changeset somehow get lost when the default branch changed from `master` to `main`? It looks like the version of `TracedError` on `main` is essentially equivalent to the version from before this PR, and I see that the PR above, targeting the v0.1.x branch, was never merged.

**Comment by BatmanAoD** [comment]
...oh, I see, `main` is a continuation of the v0.1 feature series, and `master` had already started on `v0.2` features. I've opened a backport here: https://github.com/tokio-rs/tracing/pull/3555

**Pull request merged**
